File: collectionview_customlayout_tvguide/customlayout_collectionview/ViewController.m
//
//  ViewController.m
//  collectionview
//
//  Created by Rajeeva Ranjan on 10/02/17.
//  Copyright © 2017 Rajeeva Ranjan. All rights reserved.
//

#import "ViewController.h"
#import "constants.h"
#import "CustomCollectionViewCell.h"
#import "headerfooterview.h"
#import "CustomFlowLayout.h"
#import "CustomCollectionViewLayout.h"

@implementation myCollectionViewController



@end

@implementation ViewController
@synthesize collectionView, numberItemsInsec1, numberItemsInsec2;


- (void)mymessage
{
    //NSLog(@"kjk");
}

- (void)gcAction:(UILongPressGestureRecognizer *)gc
{
    switch (gc.state) {
        case UIGestureRecognizerStateBegan:
        {
            NSLog(@"started with point %@", NSStringFromCGPoint([gc locationInView:collectionView]));
            CGPoint point = [gc locationInView:collectionView];
            NSIndexPath *selectedIndex = [collectionView indexPathForItemAtPoint:point];
            if (!selectedIndex) {
                break;
            }
            [collectionView beginInteractiveMovementForItemAtIndexPath:selectedIndex];
            break;
        }
        case UIGestureRecognizerStateChanged:
        {
            NSLog(@"chnaged with point %@", NSStringFromCGPoint([gc locationInView:collectionView]));
            CGPoint point = [gc locationInView:collectionView];
            
            [collectionView updateInteractiveMovementTargetPosition:point];
            break;
        }
            
        case UIGestureRecognizerStateEnded:
        {
            NSLog(@"ended with point %@", NSStringFromCGPoint([gc locationInView:collectionView]));

            [collectionView endInteractiveMovement];
            break;
        }
        default:
            
            
        [collectionView cancelInteractiveMovement];
    }
    
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    numberItemsInsec1 = NUMBER_OF_ITEMS_IN_SEC1;
    numberItemsInsec2 = NUMBER_OF_ITEMS_IN_SEC2;
    
    
    UICollectionViewFlowLayout *flowLayout = [[UICollectionViewFlowLayout alloc] init];
    
    
    collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 20, CV_WIDTH, CV_HEIGHT) collectionViewLayout:flowLayout];
    //collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:flowLayout];
    //collectionView.bounces = NO;
    
    UILongPressGestureRecognizer *gc = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(gcAction:)];
    //gc.allowableMovement = 500;
    
    [collectionView addGestureRecognizer:gc];
    
    
    //UICollectionViewController *myCVC = [[UICollectionViewController alloc] initWithCollectionViewLayout:flowLayout];
    //myCVC.collectionView = collectionView;

    //collectionView.allowsMultipleSelection = YES;
    
    UIView *bgView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, CV_WIDTH, CV_HEIGHT)];
    bgView.backgroundColor = [UIColor cyanColor];
    collectionView.backgroundView = bgView;
    collectionView.directionalLockEnabled = YES;
    //collectionView.multipleTouchEnabled = YES;
    
 //   [self setVerticalLayoutAttribtes];

    [self setCustomLayoutAttibutes];
    
    [collectionView setDataSource:self];
    [collectionView setDelegate:self];
    [collectionView registerClass:[CustomCollectionViewCell class] forCellWithReuseIdentifier:@"cellIdentifier"];
    [collectionView registerClass:[headerfooterview class] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"header"];
    [collectionView registerClass:[headerfooterview class] forSupplementaryViewOfKind:UICollectionElementKindSectionFooter withReuseIdentifier:@"footer"];
    
    
    [collectionView setBackgroundColor:[UIColor whiteColor]];
    
    [self.view addSubview:collectionView];
    
    
    UIButton *b1 = [UIButton buttonWithType:UIButtonTypeSystem];
    b1.frame = CGRectMake(0, DEVICE_HEIGHT -50, 130, 50);
    [b1 setTitle:@"Vertical Layout" forState:0];
    [b1 setTitle:@"Vertical Layout" forState:1];
    [b1 addTarget:self action:@selector(actionV) forControlEvents:UIControlEventTouchDown];
    [self.view addSubview:b1];
    
    UIButton *b2 = [UIButton buttonWithType:UIButtonTypeSystem];
    b2.frame = CGRectMake(130, DEVICE_HEIGHT -50, 150, 50);
    [b2 setTitle:@"Horizontal Layout" forState:0];
    [b2 setTitle:@"Horizontal Layout" forState:1];
    [b2 addTarget:self action:@selector(actionH) forControlEvents:UIControlEventTouchDown];
    [self.view addSubview:b2];
    
    UIButton *b3 = [UIButton buttonWithType:UIButtonTypeSystem];
    b3.frame = CGRectMake(280, DEVICE_HEIGHT -50, 130, 50);
    [b3 setTitle:@"Custom Layout" forState:0];
    [b3 setTitle:@"Custom Layout" forState:1];
    [b3 addTarget:self action:@selector(actionC) forControlEvents:UIControlEventTouchDown];
    [self.view addSubview:b3];
    
    
    UIButton *b4 = [UIButton buttonWithType:UIButtonTypeSystem];
    b4.frame = CGRectMake(0, DEVICE_HEIGHT -100, 130, 50);
    [b4 setTitle:@"Add at one" forState:0];
    [b4 setTitle:@"Add at one" forState:1];
    [b4 addTarget:self action:@selector(addAtOne) forControlEvents:UIControlEventTouchDown];
    [self.view addSubview:b4];
    
    
    UIButton *b5 = [UIButton buttonWithType:UIButtonTypeSystem];
    b5.frame = CGRectMake(150, DEVICE_HEIGHT -100, 130, 50);
    [b5 setTitle:@"Remove at one" forState:0];
    [b5 setTitle:@"Remove at one" forState:1];
    [b5 addTarget:self action:@selector(removeFromOne) forControlEvents:UIControlEventTouchDown];
    [self.view addSubview:b5];

    // Do any additional setup after loading the view, typically from a nib.
}

#pragma mark ADD/REMOVE apis

- (void)addAtOne
{
    numberItemsInsec1++;
    [collectionView insertItemsAtIndexPaths:[NSArray arrayWithObject:[NSIndexPath indexPathForItem:1 inSection:0]]];
}


- (void)removeFromOne
{
    numberItemsInsec1--;
    [collectionView deleteItemsAtIndexPaths:[NSArray arrayWithObject:[NSIndexPath indexPathForItem:1 inSection:0]]];

}

#pragma mark LAYOUT TYPE SETUP

- (void)setVerticalLayoutAttribtes
{
    UICollectionViewFlowLayout *flowLayout = [[UICollectionViewFlowLayout alloc] init];
    flowLayout.minimumLineSpacing = 1;
    
    flowLayout.minimumInteritemSpacing = 1;
    flowLayout.itemSize = CGSizeMake(ITEM_WIDTH, ITEM_HEIGHT);
    //flowLayout.estimatedItemSize = CGSizeMake(ITEM_WIDTH, ITEM_HEIGHT);
    
    flowLayout.headerReferenceSize = CGSizeMake(50, 50);
    flowLayout.footerReferenceSize = CGSizeMake(100, 100);
    [collectionView setCollectionViewLayout:flowLayout animated:YES];
}

- (void)setHorizontalLayoutAttribtes
{
    UICollectionViewFlowLayout *flowLayout = [[UICollectionViewFlowLayout alloc] init];
    flowLayout.minimumLineSpacing = 1;
    flowLayout.minimumInteritemSpacing = 1;
    //flowLayout.estimatedItemSize = CGSizeMake(ITEM_WIDTH, ITEM_HEIGHT);
    
    flowLayout.headerReferenceSize = CGSizeMake(50, 50);
    flowLayout.footerReferenceSize = CGSizeMake(100, 0100);
    
    [collectionView setCollectionViewLayout:flowLayout animated:YES];
    
    flowLayout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
}

- (void)setCustomLayoutAttibutes
{
    CustomCollectionViewLayout *layout = [[CustomCollectionViewLayout alloc] init];
    [collectionView setCollectionViewLayout:layout animated:YES];
    numberItemsInsec1 = NUMBER_OF_TIME_SLOTS;
    numberItemsInsec2 = NUMBER_OF_TIME_SLOTS;
    [collectionView reloadData];
    
}

- (void)actionV
{
    [self setVerticalLayoutAttribtes];
}

- (void)actionH
{
//    [collectionView reloadData];
//    CGSize size = [collectionView sizeThatFits:CGSizeMake(300, 5000)];
//    collectionView.frame = CGRectMake(0, 20, size.width, size.height);
    
    [self setHorizontalLayoutAttribtes];
}

- (void)actionC
{
    [self setCustomLayoutAttibutes];
}


#pragma mark COLLECTION VIEW SETUP

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    if (section == 0) {
        return numberItemsInsec1;
    } else {
        return numberItemsInsec2;
    }
    return numberItemsInsec1;
}

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return NUMBER_OF_SECTIONS;
}

- (void)collectionView:(UICollectionView *)collectionView willDisplayCell:(UICollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath
{
/*    [UIView beginAnimations:nil context:nil];
    [UIView setAnimationDuration:1];
    //[UIView setAnimationDelegate:cell];
    cell.layer.opacity = 0.6;
    [UIView commitAnimations];
*/
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView1 viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
{
    NSLog(@"index path %@", indexPath);
    UICollectionReusableView *view = nil;
    if ([kind isEqualToString:UICollectionElementKindSectionHeader]) {
        view = [collectionView1 dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:@"header" forIndexPath:indexPath];
        view.backgroundColor = [UIColor redColor];
    } else {
        view = [collectionView1 dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:@"footer" forIndexPath:indexPath];
        view.backgroundColor = [UIColor greenColor];
    }
    return view;
}


//First item, width and height = 20%
//2nd item, width & height = 30 %
//3rd item, width & height = 40 %
//
//4th item, width & height = 20%
//5th item, width & height = 20%
//6th item, width & height = 20%
//7th item, width & height = 20%
//
//8th item, width & height = 50%


- (CGSize)collectionView:(UICollectionView *)collectionView
                  layout:(UICollectionViewLayout *)collectionViewLayout
  sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    CGSize size = CGSizeZero;
    if (indexPath.item == 0) {
        size = CGSizeMake(CV_WIDTH*.2, CV_WIDTH*.2);//CGSizeMake(200, 50);
    } else if (indexPath.item == 1) {
        size = CGSizeMake(CV_WIDTH*.3, CV_WIDTH*.3);//CGSizeMake(100, 100);
    } else if (indexPath.item == 2) {
        size = CGSizeMake(CV_WIDTH*.4, CV_WIDTH*.4);//CGSizeMake(50, 200);
    } else if (indexPath.item == 3){
        size = CGSizeMake(CV_WIDTH*.2, CV_WIDTH*.2);//CGSizeMake(270, 100);
    } else if (indexPath.item == 4) {
        size = CGSizeMake(CV_WIDTH*.2, CV_WIDTH*.2);//CGSizeMake(100, 100);
    } else if (indexPath.item == 5) {
        size = CGSizeMake(CV_WIDTH*.2, CV_WIDTH*.2);//CGSizeMake(50, 200);
    } else if (indexPath.item == 6) {
        size = CGSizeMake(CV_WIDTH*.2, CV_WIDTH*.2);//CGSizeMake(270, 100);
    } else if (indexPath.item == 7) {
        size = CGSizeMake(CV_WIDTH*.5, CV_WIDTH*.5);//CGSizeMake(270, 100);
        
    } else {
        size = CGSizeMake(ITEM_WIDTH, ITEM_HEIGHT);
    }
    return size;
}

//- (void)collectionView:(UICollectionView *)collectionView1 didSelectItemAtIndexPath:(NSIndexPath *)indexPath
//{
//    NSLog(@"cell selected");
//}

// The cell that is returned must be retrieved from a call to -dequeueReusableCellWithReuseIdentifier:forIndexPath:
- (CustomCollectionViewCell *)collectionView:(UICollectionView *)collectionView1 cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    CustomCollectionViewCell *cell = [collectionView1 dequeueReusableCellWithReuseIdentifier:@"cellIdentifier" forIndexPath:indexPath];
    cell.clipsToBounds = YES;
    cell.backgroundColor = [UIColor clearColor];
    cell.layer.borderWidth = 1;
    
    UIView *bv = [[UIView alloc] initWithFrame:cell.frame];
    bv.backgroundColor = [UIColor lightGrayColor];
    UIView *sbv = [[UIView alloc] initWithFrame:cell.frame];
    sbv.backgroundColor = [UIColor purpleColor];
    cell.backgroundView = bv;
    cell.selectedBackgroundView = sbv;
    
    
    if (!cell.myLabel) {
        //cell.myLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, cell.frame.size.width, cell.frame.size.height)];
        cell.myLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
        cell.myLabel.opaque = YES;
        //cell.myLabel.lineBreakMode = NSLineBreakByTruncatingTail;

        [cell.myLabel setTextColor:[UIColor blackColor]];
        cell.myLabel.textAlignment = NSTextAlignmentCenter;
        [cell.contentView addSubview:cell.myLabel];
    
    }
    
    if (indexPath.section == 4) {
        if (indexPath.item == 0) {
            cell.myLabel.frame = CGRectMake(0, 0, CV_WIDTH, 40);
            cell.myLabel.text = @"ADVERTISEMENT";
        } else {
            cell.myLabel.frame = CGRectMake(0, 0, 40, 40);
            cell.myLabel.text = nil;
            
        }
    } else if (indexPath.section) {
        cell.myLabel.text = [[@"p" stringByAppendingString:[NSString stringWithFormat:@"%ld", (long)indexPath.section]] stringByAppendingString:[NSString stringWithFormat:@"%ld", (long)indexPath.item+1]];
    } else {
        cell.myLabel.text = [@"" stringByAppendingString:[NSString stringWithFormat:@"%ld", (long)indexPath.item+1]];
    }
    
    return cell;
}

#pragma mark COPY PASTE


//- (BOOL)collectionView:(UICollectionView *)collectionView shouldShowMenuForItemAtIndexPath:(NSIndexPath *)indexPath
//{
//    return YES;
//}
//
//static int itemcopied = 0;
//
//- (BOOL)collectionView:(UICollectionView *)collectionView canPerformAction:(SEL)action forItemAtIndexPath:(NSIndexPath *)indexPath withSender:(nullable id)sender
//{
//    return YES;
//}
//- (void)collectionView:(UICollectionView *)collectionView performAction:(SEL)action forItemAtIndexPath:(NSIndexPath *)indexPath withSender:(nullable id)sender
//{
//    if ([NSStringFromSelector(action) isEqualToString:@"cut:"]) {
//        [self removeFromOne];
//    }
//    
//    if ([NSStringFromSelector(action) isEqualToString:@"paste:"]) {
//        [self addAtOne];
//    }
//}

#pragma mark ITEM MOVE CODE

//- (BOOL)collectionView:(UICollectionView *)collectionView canMoveItemAtIndexPath:(NSIndexPath *)indexPath
//{
//    return YES;
//}

- (void)collectionView:(UICollectionView *)collectionView moveItemAtIndexPath:(NSIndexPath *)sourceIndexPath toIndexPath:(NSIndexPath*)destinationIndexPath
{
    
    
}



@end

File: customlayout_collectionview/ViewController.swift
//
//  ViewController.swift
//  customlayout_collectionview
//
//  Created by Rajeeva Ranjan on 21/10/20.
//  Copyright © 2020 Rajeeva Ranjan. All rights reserved.
//

import Foundation
import UIKit

class ViewController: UIViewController, UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
    
    // MARK: *** Properties ***

    var collectionview = UICollectionView.init(frame: CGRect(x: 0, y: 0, width: CV_WIDTH, height: CV_HEIGHT), collectionViewLayout: CustomCollectionViewLayout())
    var numberItemsInsec1: NSInteger = 0
    var numberItemsInsec2: NSInteger = 0
    var DEVICE_HEIGHT = UIScreen.main.bounds.size.height
    var DEVICE_WIDTH = UIScreen.main.bounds.size.width
    var b1: UIButton = UIButton()
    var b2: UIButton = UIButton()
    var b3: UIButton = UIButton()
    var b4: UIButton = UIButton()
    var b5: UIButton = UIButton()
    
    
    // MARK: *** View Controller apis ***

    override func viewDidLoad() {
        
        super.viewDidLoad()
        numberItemsInsec1 = NSInteger(NUMBER_OF_ITEMS_IN_SEC1)
        numberItemsInsec2 = NSInteger(NUMBER_OF_ITEMS_IN_SEC2)
        
        let gc = UILongPressGestureRecognizer(target: self, action: #selector(gcAction(gc:)))
        collectionview.addGestureRecognizer(gc)
        
        let bgView = UIView.init(frame: CGRect(x: 0, y: 0, width: CV_WIDTH, height: CV_HEIGHT))
        bgView.backgroundColor = UIColor.cyan
        collectionview.isDirectionalLockEnabled = true
        collectionview.backgroundView = bgView
        collectionview.bounces = false
            
        collectionview.dataSource = self
        collectionview.delegate = self
        collectionview.register(CustomCollectionViewCell.self, forCellWithReuseIdentifier: "cellIdentifier")
        collectionview.register(headerfooterview.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "header")
        collectionview.register(headerfooterview.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionFooter, withReuseIdentifier: "footer")
        
        collectionview.backgroundColor = UIColor.white
        self.view.addSubview(collectionview)
        
        b1 = UIButton.init(type: UIButton.ButtonType.system)
        b1.frame = CGRect(x: 0, y: DEVICE_HEIGHT-75, width: DEVICE_WIDTH/3, height: 50)
        b1.setTitle("Vertical Layout", for: UIControl.State.normal)
        b1.setTitle("Vertical Layout", for: UIControl.State.focused)
        b1.addTarget(self, action: #selector(setVerticalLayoutAttributes), for: UIControl.Event.touchDown)

        b2 = UIButton.init(type: UIButton.ButtonType.system)
        b2.frame = CGRect(x: DEVICE_WIDTH/3, y: DEVICE_HEIGHT-75, width: DEVICE_WIDTH/3, height: 50)
        b2.setTitle("Horizontal Layout", for: UIControl.State.normal)
        b2.setTitle("Horizontal Layout", for: UIControl.State.focused)
        b2.addTarget(self, action: #selector(setHorizontalLayoutAttributes), for: UIControl.Event.touchDown)
        
        b3 = UIButton.init(type: UIButton.ButtonType.system)
        b3.frame = CGRect(x: DEVICE_WIDTH*2/3, y: DEVICE_HEIGHT-75, width: DEVICE_WIDTH/3, height: 50)
        b3.setTitle("Custom Layout", for: UIControl.State.normal)
        b3.setTitle("Custom Layout", for: UIControl.State.focused)
        b3.addTarget(self, action: #selector(setCustomLayoutAttributes), for: UIControl.Event.touchDown)
        
        b4 = UIButton.init(type: UIButton.ButtonType.system)
        b4.frame = CGRect(x: 0, y: DEVICE_HEIGHT-125, width: DEVICE_WIDTH/2, height: 50)
        b4.setTitle("Add item at 2nd index", for: UIControl.State.normal)
        b4.setTitle("Add item at 2nd index", for: UIControl.State.focused)
        b4.addTarget(self, action: #selector(addAtOne), for: UIControl.Event.touchDown)
        
        b5 = UIButton.init(type: UIButton.ButtonType.system)
        b5.frame = CGRect(x: DEVICE_WIDTH/2, y: DEVICE_HEIGHT-125, width: DEVICE_WIDTH/2, height: 50)
        b5.setTitle("Remove item at 2nd index", for: UIControl.State.normal)
        b5.setTitle("Remove item at 2nd index", for: UIControl.State.focused)
        b5.addTarget(self, action: #selector(removeAtOne), for: UIControl.Event.touchDown)
        
        b1.backgroundColor = UIColor.systemBlue
        b1.setTitleColor(UIColor.white, for: UIControl.State.normal)
        b1.setTitleColor(UIColor.white, for: UIControl.State.focused)
        b1.layer.cornerRadius = 10

        b2.backgroundColor = UIColor.systemBlue
        b2.setTitleColor(UIColor.white, for: UIControl.State.normal)
        b2.setTitleColor(UIColor.white, for: UIControl.State.focused)
        b2.layer.cornerRadius = 10
        
        b3.backgroundColor = UIColor.systemBlue
        b3.setTitleColor(UIColor.white, for: UIControl.State.normal)
        b3.setTitleColor(UIColor.white, for: UIControl.State.focused)
        b3.layer.cornerRadius = 10
        
        b4.backgroundColor = UIColor.systemBlue
        b4.setTitleColor(UIColor.white, for: UIControl.State.normal)
        b4.setTitleColor(UIColor.white, for: UIControl.State.focused)
        b4.layer.cornerRadius = 10
        
        b5.backgroundColor = UIColor.systemBlue
        b5.setTitleColor(UIColor.white, for: UIControl.State.normal)
        b5.setTitleColor(UIColor.white, for: UIControl.State.focused)
        b5.layer.cornerRadius = 10

        self.view.addSubview(b1)
        self.view.addSubview(b2)
        self.view.addSubview(b3)
        self.view.addSubview(b4)
        self.view.addSubview(b5)
        
        self.setCustomLayoutAttributes()
        //collectionview.reloadData();

    }
    
    
    override func viewDidAppear(_ animated: Bool) {
        var collViewFrame = collectionview.frame
        collViewFrame = CGRect(x: collViewFrame.origin.x, y: 0, width: collViewFrame.size.width, height: DEVICE_HEIGHT)
        collectionview.contentInsetAdjustmentBehavior = .always
        collectionview.frame = collViewFrame

    }
    
    // MARK: CollectionView datasource
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if (section == 0) {
            return numberItemsInsec1;
        }
        
        return numberItemsInsec2;
    }
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return Int(NUMBER_OF_SECTIONS)
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell1 = collectionview.dequeueReusableCell(withReuseIdentifier: "cellIdentifier", for: indexPath)
        
        let cell = cell1 as! CustomCollectionViewCell
        
        cell.clipsToBounds = true
        cell.backgroundColor = UIColor.clear
        cell.layer.borderWidth = 1
        
        let bv = UIView.init(frame: cell.frame)
        bv.backgroundColor = UIColor.cyan
        
        let sbv = UIView.init(frame: cell.frame)
        sbv.backgroundColor = UIColor.systemBlue
        
        cell.backgroundView = bv;
        cell.selectedBackgroundView = sbv;
                
        if (cell.myLabel == nil) {

            cell.myLabel = UILabel.init(frame: CGRect(x: 0, y: 0, width: 150, height: 75))
            cell.myLabel.isOpaque = true;

//            cell.myLabel.textColor = UIColor.black
//            cell.myLabel.textAlignment = NSTextAlignment.center
            cell.contentView.addSubview(cell.myLabel)
        
        }
        
        cell.myLabel.textColor = UIColor.black
        cell.myLabel.textAlignment = NSTextAlignment.left
        cell.myLabel.backgroundColor = UIColor.clear

        
        if !collectionview.collectionViewLayout.isMember(of: CustomCollectionViewLayout.self) {
           
            cell.myLabel.text = "Item ".appendingFormat("%d", indexPath.section).appendingFormat("%d", indexPath.item+1)
            cell.myLabel.frame = CGRect(x: 0, y: 0, width: 150, height: 75)


        } else {
            if (indexPath.section == 4) {
                if (indexPath.item == 0) {
                    cell.myLabel.frame = CGRect(x: 0, y: 0, width: CV_WIDTH, height: 75)
                    cell.myLabel.text = "COMMERCIAL";
                    cell.myLabel.textAlignment = NSTextAlignment.center
                    cell.myLabel.backgroundColor = UIColor.red
                    cell.myLabel.textColor = UIColor.yellow

                } else {
                    cell.myLabel.frame = CGRect(x: 0, y: 0, width: 150, height: 75)
                    cell.myLabel.text = nil;
                    
                }
            } else if (indexPath.section != 0) {
                cell.myLabel.text = " Prog ".appendingFormat("%d", indexPath.section).appendingFormat("%d", indexPath.item+1)
            } else {
                cell.myLabel.text = " Time slot ".appendingFormat("%d", indexPath.section).appendingFormat("%d", indexPath.item+1)
                cell.myLabel.backgroundColor = UIColor.systemBlue
                cell.myLabel.textColor = UIColor.white
                cell.myLabel.textAlignment = NSTextAlignment.center


                
            }
        }
        
        return cell;
    }
    
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        
        var view: UICollectionReusableView;
        if (kind == UICollectionView.elementKindSectionHeader) {
            view = collectionview.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: "header", for: indexPath)
            view.backgroundColor = UIColor.systemBlue;
        } else {
            view = collectionview.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: "footer", for: indexPath)
            view.backgroundColor = UIColor.purple;
        }
        return view;
    }
    
    // MARK: *** Data operation ***
    
    @objc func addAtOne() {
        numberItemsInsec1 += 1
        collectionview.insertItems(at: [IndexPath(item: 1, section: 0)])
    }
    
    @objc func removeAtOne() {
        numberItemsInsec1 -= 1
        collectionview.deleteItems(at: [IndexPath(item: 1, section: 0)])
    }
    
    // MARK: *** Layout ***
        
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        var size = CGSize(width: 0, height: 0);
        if (indexPath.item == 0) {
            size = CGSize(width: CV_WIDTH*0.2, height: CV_WIDTH*0.2)
        } else if (indexPath.item == 1) {
            size = CGSize(width: CV_WIDTH*0.3, height: CV_WIDTH*0.3)
        } else if (indexPath.item == 2) {
            size = CGSize(width: CV_WIDTH*0.4, height: CV_WIDTH*0.4)
        } else if (indexPath.item == 3){
            size = CGSize(width: CV_WIDTH*0.2, height: CV_WIDTH*0.2)
        } else if (indexPath.item == 4) {
            size = CGSize(width: CV_WIDTH*0.2, height: CV_WIDTH*0.2)
        } else if (indexPath.item == 5) {
            size = CGSize(width: CV_WIDTH*0.2, height: CV_WIDTH*0.2)
        } else if (indexPath.item == 6) {
            size = CGSize(width: CV_WIDTH*0.2, height: CV_WIDTH*0.2)
        } else if (indexPath.item == 7) {
            size = CGSize(width: CV_WIDTH*0.5, height: CV_WIDTH*0.5)
        } else {
            size = CGSize(width: Int(ITEM_WIDTH), height: Int(ITEM_HEIGHT))
        }
        return size;
    }
    
    
    // MARK: *** Item movement ***

    func collectionView(_ collectionView: UICollectionView, canMoveItemAt indexPath: IndexPath) -> Bool {
        if !collectionview.collectionViewLayout.isMember(of: CustomCollectionViewLayout.self) {
            return true
        }
        return false
    }

    func collectionView(_ collectionView: UICollectionView, moveItemAt sourceIndexPath: IndexPath, to destinationIndexPath: IndexPath) {

    }
    
    @objc func gcAction(gc: UILongPressGestureRecognizer) {
        
        switch gc.state {
        
        case .began:

            let point: CGPoint = gc.location(in: collectionview)
            let indexPath = collectionview.indexPathForItem(at: point)
            if indexPath == nil {
                break
            }
            collectionview.beginInteractiveMovementForItem(at: indexPath!)
            
        case .changed:
 
            let point = gc.location(in: collectionview)
            collectionview.updateInteractiveMovementTargetPosition(point)
        
        case .ended:
            collectionview.endInteractiveMovement()

        default:
            collectionview.cancelInteractiveMovement()
        }
    }
    
    // MARK: *** Util methods ***
    
    func createTemplateFlowLayout() -> UICollectionViewFlowLayout {
        let flowLayout = UICollectionViewFlowLayout.init()
        flowLayout.minimumLineSpacing = 1
        flowLayout.minimumInteritemSpacing = 1
        flowLayout.itemSize = CGSize(width: Int(ITEM_WIDTH), height: Int(ITEM_HEIGHT))
        flowLayout.headerReferenceSize = CGSize(width: 50, height: 50)
        flowLayout.footerReferenceSize = CGSize(width: 50, height: 50)
        return flowLayout
    }
    
    @objc func setVerticalLayoutAttributes() {
        
        let flowLayout = createTemplateFlowLayout()
        //collectionview.collectionViewLayout.invalidateLayout()

        collectionview.setCollectionViewLayout(flowLayout, animated: false)
        collectionview.reloadData();
        
        flowLayout.scrollDirection = UICollectionView.ScrollDirection.horizontal
        
        //collectionview.setContentOffset(CGPoint(x: 0, y: 0), animated: true)

        b4.isHidden = false
        b5.isHidden = false
    }
    
    @objc func setHorizontalLayoutAttributes() {

        let flowLayout = createTemplateFlowLayout()
        //collectionview.collectionViewLayout.invalidateLayout()

        collectionview.setCollectionViewLayout(flowLayout, animated: false)
        collectionview.reloadData();
        
        flowLayout.scrollDirection = UICollectionView.ScrollDirection.vertical
                
        //collectionview.setContentOffset(CGPoint(x: 0, y: 0), animated: true)


        b4.isHidden = false
        b5.isHidden = false
    }
    
    @objc func setCustomLayoutAttributes() {
        
        let layout = CustomCollectionViewLayout()
        //collectionview.collectionViewLayout.invalidateLayout()
        collectionview.setCollectionViewLayout(layout, animated: false)
        numberItemsInsec1 = NSInteger(NUMBER_OF_TIME_SLOTS);
        numberItemsInsec2 = NSInteger(NUMBER_OF_TIME_SLOTS);
        
        collectionview.reloadData()
        
        //collectionview.setContentOffset(CGPoint(x: 0, y: 0), animated: true)
        
        b4.isHidden = true
        b5.isHidden = true
    }

}
